File: 23. Intersection point of two lists.cpp
#include <iostream>
using namespace std; struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};
int getCount(Node* head){ int c=0; while(head){ c++; head=head->next; } return c; }
Node* getIntersection(Node* a, Node* b){ int c1=getCount(a); int c2=getCount(b); int d=abs(c1-c2); if(c1>c2){ for(int i=0;i<d;i++) a=a->next; } else { for(int i=0;i<d;i++) b=b->next; } while(a && b){ if(a==b) return a; a=a->next; b=b->next; } return nullptr; }
int main(){ Node* common=new Node(8); common->next=new Node(10);
    Node* a=new Node(3); a->next=new Node(6); a->next->next=common;
    Node* b=new Node(4); b->next=common;
    Node* inter=getIntersection(a,b);
    if(inter) cout<<"Intersection at: "<<inter->data<<endl; else cout<<"No intersection"<<endl;
    return 0; }

File: 23. Intersection point of two lists.py
class Node:
    def __init__(self,data): self.data=data; self.next=None

def get_count(head):
    c=0
    while head:
        c+=1; head=head.next
    return c

def get_intersection(a,b):
    c1=get_count(a); c2=get_count(b)
    d=abs(c1-c2)
    if c1>c2:
        for _ in range(d): a=a.next
    else:
        for _ in range(d): b=b.next
    while a and b:
        if a is b: return a
        a=a.next; b=b.next
    return None

if __name__=='__main__':
    common=Node(8); common.next=Node(10)
    a=Node(3); a.next=Node(6); a.next.next=common
    b=Node(4); b.next=common
    inter=get_intersection(a,b)
    print('Intersection at:', inter.data if inter else None)
